From 2775cdefdca8cdaa5ada7aeeb936bb17bf893faf Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Tue, 29 Oct 2013 09:51:40 +0700
Subject: Dalvik: RegisterMap cannot contain more then 65536 GC points

BZ: 143225

RegisterMap contains an address (offset) of the GC point in a method.
For this purpose it has no more than 16 bit as a result offset cannot
exceed 65536. However the implementation does not check this condition
resulting in incorrect behavior.

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: yes

Change-Id: I26a7cb988b6add168828d886c56ed4567cce0dcb
Orig-MCG-Change-Id: If8482f3ad13ff6561102ce020f1433c91097c1e9
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/analysis/RegisterMap.cpp |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/vm/analysis/RegisterMap.cpp b/vm/analysis/RegisterMap.cpp
index 197fb7a..82f946b 100644
--- a/vm/analysis/RegisterMap.cpp
+++ b/vm/analysis/RegisterMap.cpp
@@ -226,8 +226,13 @@ RegisterMap* dvmGenerateRegisterMapV(VerifierData* vdata)
      */
     gcPointCount = 0;
     for (i = 0; i < (int) vdata->insnsSize; i++) {
-        if (dvmInsnIsGcPoint(vdata->insnFlags, i))
+        if (dvmInsnIsGcPoint(vdata->insnFlags, i)) {
+            if (i >= 65536) {
+                ALOGE("ERROR: register map can't handle %d instructions in one method", i);
+                goto bail;
+            }
             gcPointCount++;
+        }
     }
     if (gcPointCount >= 65536) {
         /* we could handle this, but in practice we don't get near this */
-- 
1.7.4.1

